subsys/bluetooth/host/scan.c

```diff
@@ -765,154 +867,266 @@ static void create_ext_adv_info(struct bt_hci_evt_le_ext_advertising_info const
+		 * from the fragmented advertising reports. If the chain is
+		 * borked, we don't bother with allocation as we will discard
+		 * the chain anyways.
+		 */
+		init_reassembly(r, &evt->addr, evt->sid);
+	}
 
-		/* No more data coming from the controller.
-		 * Create event.
+	if (r->report == NULL) {
+		LOG_DBG("No reassembly buf.");
+		start_discarding(r);
+	} else if (evt->length > net_buf_tailroom(r->report)) {
+		LOG_DBG("Adv report doesn't fit reassembly buffer.");
+		start_discarding(r);
+	}
+
+	bool discarding = r->state == FRAG_ADV_DISCARDING;
+
+	if (!discarding) {
+		__ASSERT_NO_MSG(r->report);
+		net_buf_add_mem(r->report, /* flex the member 🤖💪 */ evt->data, evt->length);
+	}
+
+	if (!discarding && complete) {
+		/* Core Spec v5.4. 4.E.7.7.65.13 says controller uses latest
+		 * RSSI. Host wants to be like controller when grow up.
 		 */
-		__ASSERT_NO_MSG(is_report_complete);
+		struct bt_le_scan_recv_info scan_info;
+
 		create_ext_adv_info(evt, &scan_info);
-		le_adv_recv(&evt->addr, &scan_info, &ext_scan_buf, ext_scan_buf.len);
+		process_reassembled_report_async(r->report, &r->addr, &scan_info);
+	}
 
-		/* We do no longer need to keep track of this advertiser. */
-		reset_reassembling_advertiser();
+	if (end_of_chain) {
+		reset_reassembly(r);
+	}
 
-cont:
-		net_buf_pull(buf, evt->length);
+	return true;
+}
+
+static bool is_part_of_buf(void *_data, struct net_buf *buf)
+{
+	uintptr_t data = POINTER_TO_UINT(_data);
+	uintptr_t start = POINTER_TO_UINT(buf->__buf);
+	uintptr_t end = POINTER_TO_UINT(buf->__buf + buf->size);
+
+	return (data >= start) && (data <= end);
+}
+
+static void process_unfragmented_report(struct bt_hci_evt_le_ext_advertising_info *evt,
+					struct net_buf *report)
+{
+	if (report == NULL) {
+#if defined(CONFIG_BT_HTOP)
+		atomic_inc(&dropped_adv_reports);
+#endif
+		return;
+	}
+
+	if (evt->length > net_buf_tailroom(report)) {
+		net_buf_unref(report);
+		report = NULL;
+		return;
+	}
+
+	if (is_part_of_buf(evt, report)) {
+		/* Oh my god karen, you can't just ask a buf to copy to itself */
+		net_buf_push(report, sizeof(*evt));
+		__ASSERT_NO_MSG((void *)(evt) == (void *)(report->data));
+	} else {
+		net_buf_reserve(report, REPORT_HEADER_SIZE);
+		/* slightly unsafe */
+		net_buf_add_mem(report, evt, sizeof(*evt) + evt->length);
+	}
+
+	net_buf_push_u8(report, COMPLETE_REPORT_MAGIC);
+
+	LOG_DBG("queue %p", report);
+
+	k_fifo_put(&adv_reports, report);
+	k_work_submit_to_queue(bt_get_rx_wq(), &adv_report_work);
+}
+
+bool bt_buf_can_steal_ext_adv_report_buf(struct net_buf *buf);
+
+static struct net_buf *get_me_a_buf(uint8_t remaining_reports, struct net_buf **buf)
+{
+	struct net_buf *b;
+
+	if (remaining_reports == 1 && bt_buf_can_steal_ext_adv_report_buf(*buf)) {
+		/* optimization, we can zero-copy instead of malloc */
+		b = net_buf_ref(*buf);
+
+		/* The caller should _not_ modify `buf` anymore. We stole it. */
+		*buf = NULL;
+	} else {
+		b = net_buf_alloc(&reassembly_pool, K_NO_WAIT);
+	}
```

Comment:
I think I'm missing some detail. What the does the number of reports have to do with reassembly? Why can't we zero-copy when `remaining_reports = 2`?

Reply:
If you were in oslo I could show you my sketches.
Basically, the optimization becomes more involved if we don't only apply it to the last report. We can only steal the `buf` from the HCI driver once. 
So if we want to steal `buf` for report number=1 out of num_reports=2, we have to:
- read from a `buf_simple` or something (and then fn signatures get really complicated)
- somehow ref `buf`, keep some state that we want to schedule it later
- read report 2, alloc and queue it
- remember we wanted to schedule report 1, fast-forward the `buf->data` pointer to report 1's location (which might not be the 1st report in real life)
- add the magic to report 1, and queue it

So it's quite a bit more code and control flow, and I'm unconvinced that we gain much from it. I think in 80% of cases we don't batch reports anyways and we'll have lots of `num_reports=1` with `COMPLETE`. This optimization is designed only for that case. The rest can "malloc", it's easier to follow.

As I said though, "I think". It would be good to collect some HCI traces at the UPF to see if SDC or other controllers tend to batch reports or send lots of small events. Then we can revisit.

Reply:
Could I bother you or @alwa-nordic to put that explanation as a in-code comment? 

#thecodeisnotselfdocumenting

Reply:
I'll definitely do that. It's always hard to put oneself in the reader's shoes